### PR TITLE
Api pass key

### PR DESCRIPTION
**Pull request by YazeedLoonat**
## Issue Overview

This PR addresses [#issue](https://github.com/bloom-housing/bloom/issues/4008)

- [x] This change addresses the issue in full

## Description
This creates a new api-pass-key guard that is applied through out the backend

## How Can This Be Tested/Reviewed?
**Add the new api .env variable**: API_PASS_KEY

if you use the site from either the public or partners sites, everything should work as normal

if you attempt to use the `localhost:3101/api` or `localhost:3100/api` sites (depending on your local set up) only 2 endpoints should be usable the `/` and the `/jurisdictions` endpoints

everything else should give you an error

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [x] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [x] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by emilyjablonski** [comment]
@YazeedLoonat Heads up that this is still in a draft state / about the failing test suites

**Comment by emilyjablonski** [comment]
@YazeedLoonat Is this ready for review? Just asking bc it has the label, but it's still in a draft state / has failing tests

File: api/.env.template [hunk 1]
```diff
@@ -50,3 +50,5 @@ TIME_ZONE=America/Los_Angeles
 THROTTLE_TTL=3600000
 # how many requests before we throttle
 THROTTLE_LIMIT=100
+# API passkey, requests missing this will not be alllowed to progress
+API_PASS_KEY="some-key-here"
```
Comment (mcgarrye) [inline api/.env.template:54]: nit: EOF

File: api/prisma/seed-staging.ts [hunk 2]
```diff
@@ -47,8 +47,12 @@ export const stagingSeed = async (
     data: jurisdictionFactory(jurisdictionName, [UserRoleEnum.admin]),
   });
   // add another jurisdiction
+  let otherJusisName = randomNoun();
+  while (otherJusisName < jurisdictionName) {
```
Comment (mcgarrye) [inline api/prisma/seed-staging.ts:51]: What are we looking for with this comparison?

File: api/test/integration/api-key-guard.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  afterAll(async () => {
+    await prisma.$disconnect();
+    await app.close();
+  });
+
+  it('should succeed when correct header is present', async () => {
+    await request(app.getHttpServer())
+      .get('/jurisdictions')
+      .set({ passkey: process.env.API_PASS_KEY || '' })
```
Comment (mcgarrye) [inline api/test/integration/api-key-guard.e2e-spec.ts:29]: Will the `process.env` cause an issue in doorway tests?

File: sites/partners/src/pages/listings/[id]/edit.tsx [hunk 4]
```diff
@@ -64,7 +64,9 @@ export async function getServerSideProps(context: { params: Record<string, strin
   let response
 
   try {
-    response = await axios.get(`${process.env.backendApiBase}/listings/${context.params.id}`)
+    response = await axios.get(`${process.env.backendApiBase}/listings/${context.params.id}`, {
+      headers: { passkey: process.env.API_PASS_KEY },
```
Comment (mcgarrye) [inline sites/partners/src/pages/listings/[id]/edit.tsx:68]: Why is this required on this specific endpoint?

File: api/test/integration/throttle-guard.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { INestApplication } from '@nestjs/common';
+import request from 'supertest';
+import { AppModule } from '../../src/modules/app.module';
+import { PrismaService } from '../../src/services/prisma.service';
+
+describe('Throttle Guard Tests', () => {
```
Comment (mcgarrye) [inline api/test/integration/throttle-guard.e2e-spec.ts:7]: What endpoint still uses this?

File: api/test/integration/api-key-guard.e2e-spec.ts (hunk 3 above)
Comment (YazeedLoonat) [reply api/test/integration/api-key-guard.e2e-spec.ts:29]: it shouldn't, we only run into testing issues on unit tests that attempt to hit the db

this is an e2e test so the db will be present, and the api is stood up in its entirety to test e2e

File: api/test/integration/throttle-guard.e2e-spec.ts (hunk 5 above)
Comment (YazeedLoonat) [reply api/test/integration/throttle-guard.e2e-spec.ts:7]: This is a bit unfortunate, but while writing this test out I found out that we were double calling the guard. Once on every request and once through specific useGuard() calls

part of this pr pulled all those useGuards out so we just have the general one that was already in place

File: sites/partners/src/pages/listings/[id]/edit.tsx (hunk 4 above)
Comment (YazeedLoonat) [reply sites/partners/src/pages/listings/[id]/edit.tsx:68]: I'm not sure I understand the question. Almost all endpoints on the api now require the passkey is sent across. If we don't send it here then the api will respond with an error and will prevent the user from making changes to listings

**Comment by mcgarrye** [comment]
`yarn dev:all` does not work with passkey changes

**Comment by YazeedLoonat** [comment]
Hey @mcgarrye I think I got all the updates you were looking for out, lemme know if you catch anything else

**Review by mcgarrye — APPROVED** [review]
Let's Go Knicks

File: api/src/guards/api-key.guard.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+
+@Injectable()
+export class ApiKeyGuard implements CanActivate {
+  canActivate(context: ExecutionContext) {
+    const req = context.switchToHttp().getRequest();
+
+    if (
```
Comment (emilyjablonski) [inline api/src/guards/api-key.guard.ts:13]: Should we also be throwing if the pass key is not set?

File: api/test/unit/utilities/generate-single-use-code.spec.ts [hunk 7]
```diff
@@ -1,22 +1,7 @@
...
-    }).compile();
-
-    app = moduleFixture.createNestApplication();
-    await app.init();
-  });
   it('should generate mfa code of the specified length', () => {
-    expect(generateSingleUseCode().length).toEqual(
-      Number(process.env.MFA_CODE_LENGTH),
-    );
+    expect(generateSingleUseCode(5).length).toEqual(5);
```
Comment (emilyjablonski) [inline api/test/unit/utilities/generate-single-use-code.spec.ts:5]: Should we still be checking for the env var?

Comment (YazeedLoonat) [reply api/test/unit/utilities/generate-single-use-code.spec.ts:5]: I ended up running into some issues with this test (mainly flakiness, sometimes passing, sometimes failing)

so what I ended up doing was changing the `generateSingleUseCode()` to take in a numeric argument. In the code we now send `Number(process.env.MFA_CODE_LENGTH)` wherever we call `generateSingleUseCode` so no need to check this function. Just make sure the length matches what the incoming arg says the length should be

File: api/src/guards/api-key.guard.ts (hunk 6 above)
Comment (YazeedLoonat) [reply api/src/guards/api-key.guard.ts:13]: It'd probably be best to keep it optional, especially on doorway there will probably be a window where the env variable hasn't been set yet and we'll want to allow requests through while that is set up/spinning up

**Review by ColinBuyck — APPROVED** [review]
LGTM with just one caveat ❇️ I wasn't able to get mfa with phone working but I so rarely test that path. Unsure if my keys are outdated or if it hasn't worked on local but could you just verify that that is working as expected on your end?

**Comment by YazeedLoonat** [comment]
Hey @ColinBuyck it seems that there is a problem on Twilio's side of things: https://status.twilio.com/

Playing around with it locally it seems to be working but twilio might be having issues with actually sending the SMS I'll merge this in and we can keep track of that

**Pull request merged**